### DISCUSSION TITLE
Why does common.sql DbApiHook does not support password?

### DISCUSSION THREAD

**Discussion by bryzgaloff** [apache/airflow#33555] [General]
### Description

The current implementation of `common.sql` provider's `DbApiHook` does not pass any sort of a password to `connect`:

https://github.com/apache/airflow/blob/c9452c8b7b61c975a1097544a00ebea00501b37d/airflow/providers/common/sql/hooks/sql.py#L175-L178

And this is because `ConnectorProtocol` does not support it either:

https://github.com/apache/airflow/blob/c9452c8b7b61c975a1097544a00ebea00501b37d/airflow/providers/common/sql/hooks/sql.py#L105-L108

Is it an intended decision complying with some standards? Or just a missing argument?

I suggest changing a call to support all the Airflow's `Connection` attributes.

I am happy to submit a PR if there are no objections or fundamental reasons to not generalize the interface.

### Use case/motivation

As an author of [Airflow ClickHouse Plugin](https://github.com/bryzgaloff/airflow-clickhouse-plugin), I want to rely on the existing `DbApiHook` implementation as much as possible without a need to override `DbApiHook.get_conn` method. And if it passes all the arguments, I need to only override the `DbApiHook.connector` object to customize the behaviour.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
This is more of a discussion than issue/feature (I convereted it to discussion now).

Generally - you are supposed to override `get_conn` method - as all other DBapiHooks do. I think the main reason for this "base" implemenation is mostly historical. If you would like to attempt to do it - feel free, but there are likely subtle backwards compatibility issues. You would have to make sure that whatever you come up with will be both future- and back- compatible. The thing with "common.sql" is that you can have newer or older version of providers installed (for example rather new Hive provider and rather old common.sql provider) - which means that keeping compatibility is very important - even if it leaves some things like that behind.

I'd strongly recommend you to just follow what other providers and Hooks do.

**Reply by bryzgaloff** [reply]
Ok, thank you for a reply. Not sure if it is worth it for me right now, so I will just override `get_conn`. But let's keep the discussion open: maybe someone will make a heroic effort to make the code more flexible and also keep the backward compatibility.

Could you please suggest ideas how to check the backwards compatibility? What would you suggest checking after the changes are made? Testing all the `common.sql`-dependent providers from an official list?

Also, won't versioning fundamentally solve this issue? We may increment the minor version to make sure the other providers creator may keep using the "stable" version.

**Reply by potiuk** [reply]
> Ok, thank you for a reply. Not sure if it is worth it for me right now, so I will just override get_conn. But let's keep the discussion open: maybe someone will make a heroic effort to make the code more flexible and also keep the backward compatibility.

Generally we don't close discussions. They might die out when nobody continues them. That's expected and property of discussions.

> Could you please suggest ideas how to check the backwards compatibility? What would you suggest checking after the changes are made? Testing all the common.sql-dependent providers from an official list?

This is tricky. we have automated checks in comon.sql API by pre-commit (if you reamove or change existing API you will have to deliberately commit a change to the API). And code reviews looking for potential incompatibilities. But you can raise deprecation warnings from methods that are not supposed to be used.

Generally speaking the way to handle it is to add new features to Common.sql and then add a minimum version of common.sq in providers that depend on it. Simply sayying (I need "common.sql >=1.X.0' - this way you do not have to check compability with older versions. 

So the way to introduce such  change is to:

* deprecate old methods and make sure they are not used
* add a new feature and add min-version for all providers that use it

I cannot give the exact recipe though. It really depends on what you (or whoever wants to do it) to reson about such change.
